ethapi/api.go

```diff
@@ -1606,7 +1607,7 @@ func (s *PublicTransactionPoolAPI) GetTransactionReceipt(ctx context.Context, ha
 	// Derive the sender.
 	bigblock := new(big.Int).SetUint64(blockNumber)
 	signer := gsignercache.Wrap(types.MakeSigner(s.b.ChainConfig(), bigblock))
-	from, _ := types.Sender(signer, tx)
+	from, _ := internaltx.Sender(signer, tx)
```

Comment:
Yes, the naming would make a bit confuse. I think it'd better to handle the `internaltx.Sender` inside `types.Sender` other than handle the `types.Sender` inside `internaltx.Sender`

Reply:
`internaltx.Sender` returns addresses for internal txs. Internal txs are not signed, they are generated inside the block processing. Internal txs are currently generated by the drivermodule and those transactions talk to the NodeDriver contact. An example is `drivercall.SealEpoc` which signals that an epoch was sealed and sends metrics for validators, by which SFC contract will decide rewards for them. In a grand scheme of things, block processor and EVM contracts communicate with each other via internal transactions and contract logs, and NodeDriver contract is a middleware between other EVM contracts and block processor

Previously, all internal transactions had zero address. Now, it allows internal transactions to have any address, which gets encoded into `S` signature number. This functionality is needed by dangerous in a sense that, if implemented carelessly, it'll allow an attacker to send transaction from any address without signing them. To minimize possibility of such a bug, I didn't create new `types.Signer` for internal txs but rather created new version of `types.Sender` so that it would be obvious in which places we allow internal txs, and in which places we clearly allow only usual external txs (e.g. API, tx pool, event checkers)

Reply:
If comparing with Ethereum, then you noticed that it just increments miner's balance at the end of block execution. Without any transactions, which violates the principle that all changes (except genesis) must be done inside txs. It's okay for them because they have exactly one place where such things happen, but in Opera, due to the type of consensus, it happens way too often and we need to make those changes inside transactions to keep things in order